crypto/aes/aes_ige.c

```diff
@@ -12,18 +12,22 @@
 #include <openssl/aes.h>
 #include "aes_locl.h"
 
-#define N_WORDS (AES_BLOCK_SIZE / sizeof(unsigned long))
-typedef struct {
-    unsigned long data[N_WORDS];
-} aes_block_t;
-
 /* XXX: probably some better way to do this */
 #if defined(__i386__) || defined(__x86_64__)
 # define UNALIGNED_MEMOPS_ARE_FAST 1
 #else
 # define UNALIGNED_MEMOPS_ARE_FAST 0
 #endif
 
+#define N_WORDS (AES_BLOCK_SIZE / sizeof(unsigned long))
+typedef struct {
+    unsigned long data[N_WORDS];
+#if defined(__GNUC__) && UNALIGNED_MEMOPS_ARE_FAST
+} aes_block_t __attribute((__aligned__(1)));
```

Comment:
Out of curiosity, `info gcc` gives alignment examples as `__attribute__(aligned(N))`, so which is "more" right? Concern basically is if gcc-alikes would treat it right.

Reply:
All attribute keywords are generally available as `keyword` and `__keyword__`,
using names beginning with double underscore has the advantage that it cannot
collide with user macro names. Which is of course not an issue here, but giving
good examples is important.
I cannot really guarantee what other compilers make out of this attribute.
Regarding GCC, there are lots of places where attributes are syntactically accepted,
but silently discarded.  My first try was adding an attribute at the using statement
`*(int __attribute((__aligned__(1))) *)p = x;` but that attribute was ignored.

Reply:
I said: The attribute works only when used with a typedef.

Reply:
BTW: in crypto/ec/ecp_nistz256_table.c I see a similar syntax:

```
#if defined(__GNUC__)
__attribute((aligned(4096)))
#elif defined(_MSC_VER)
__declspec(align(4096))
#elif defined(__SUNPRO_C)
# pragma align 4096(ecp_nistz256_precomputed)
#endif
```

Reply:
It doesn't mean that it's right :-) Note that it's contributed code. Also note that it's not fatal if ignored. Well, one argue that it's not fatal here either, but as you said yourself, giving good example is important, and goodness is defined by consistency with documentation or some claimed advantage.

Reply:
BTW, not that file in question is not actually passed to compiler. Instead it's parsed by corresponding assembly module. And nistz only available in assembly build... Well, not that it's an excuse of some kind...